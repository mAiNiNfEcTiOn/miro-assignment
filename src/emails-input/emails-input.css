.emails-input__root {
  -ms-flex-line-pack: start;
  align-content: flex-start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: flex-start;
  background: #ffffff;
  border-radius: 4px;
  border: 1px solid #c3c2cf;
  cursor: text;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 0;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  overflow: auto;
  padding: 5px;
  width: 100%;
}

.emails-input__root__input {
  border: none;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-box-flex: 1;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  font-family: Open Sans;
  font-size: 14px;
  height: 24px;
  line-height: 24px;
  outline: none;
}

.emails-input__root__input::-webkit-input-placeholder {
  color: #c3c2cf;
  font-family: Open Sans;
  font-size: 14px;
  font-style: normal;
  font-weight: normal;
  line-height: 24px;
}

.emails-input__root__input::-webkit-input-placeholder,
.emails-input__root__input::-moz-placeholder,
.emails-input__root__input:-ms-input-placeholder,
.emails-input__root__input::-ms-input-placeholder,
.emails-input__root__input::placeholder {
  color: #c3c2cf;
  font-family: Open Sans;
  font-size: 14px;
  font-style: normal;
  font-weight: normal;
  line-height: 24px;
}

.emails-input__root__pill {
  background: rgba(102, 153, 255, 0.2);
  border-radius: 100px;
  color: #050038;
  cursor: default;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 0;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  font-family: Open Sans;
  font-size: 14px;
  font-weight: normal;
  height: 24px;
  line-height: 24px;
  margin-right: 8px;
  margin-top: 4px;
  padding: 0 8px 0 10px;
}

.emails-input__root__pill--invalid {
  background: #fff;
  border-bottom: 1px dashed #d92929;
  border-radius: 0;
  padding: 0;
}

.emails-input__root__pill__button {
  background: none;
  border: none;
  margin-left: 8px;
}
