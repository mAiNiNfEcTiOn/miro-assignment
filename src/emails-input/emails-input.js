import "./emails-input.css";

((window, document) => {
  const chars = [];
  for (let charCode = 48; charCode <= 57; charCode += 1) {
    chars.push(String.fromCharCode(charCode));
  }

  for (let charCode = 65; charCode <= 90; charCode += 1) {
    chars.push(String.fromCharCode(charCode));
  }

  for (let charCode = 97; charCode <= 122; charCode += 1) {
    chars.push(String.fromCharCode(charCode));
  }

  function generateRandomEmail(size = 8) {
    let email = "";
    for (let charCounter = 0; charCounter < size; charCounter += 1) {
      email += chars[Math.floor(Math.random() * chars.length)];
    }

    email += "@miro.com";

    return email;
  }

  function handleInputValue(config) {
    const value = config.value;
    const store = config.store;

    if (value) {
      value.split(",").forEach(item => {
        if (!item) {
          return;
        }

        item = item.trim();

        const alreadyExists = store.some(record => record.value === item);

        if (alreadyExists) {
          return;
        }

        const emailField = document.createElement("input");
        emailField.type = "email";
        emailField.value = item;

        const isValid = emailField.checkValidity();

        store.push({ value: item, isValid: isValid });
        config.node.insertAdjacentElement(
          "beforebegin",
          Pill({ value: item, isValid: isValid, store: store })
        );
      });
    }
  }

  /**
   * 
   * @param {*} config 
   */
  function Pill(config) {
    const containerDiv = document.createElement("div");
    const span = document.createElement("span");
    const button = document.createElement("button");

    button.className = "emails-input__root__pill__button";
    containerDiv.className = "emails-input__root__pill";

    span.textContent = config.value;
    button.type = "button";
    button.innerHTML =
      '<svg focusable="false" tabindex="-1" width="8" height="8" viewBox="0 0 8 8" fill="none" xmlns="http://www.w3.org/2000/svg">' +
      '<path fill-rule="evenodd" clip-rule="evenodd" d="M8 0.8L7.2 0L4 3.2L0.8 0L0 0.8L3.2 4L0 7.2L0.8 8L4 4.8L7.2 8L8 7.2L4.8 4L8 0.8Z" fill="#050038"/>' +
      "</svg>";

    button.addEventListener("click", e => {
      const pill = button.parentNode;
      for (let idx = 0; idx < config.store.length; idx += 1) {
        if (config.store[idx].value === config.value) {
          config.store.splice(idx, 1);
          break;
        }
      }

      pill.parentNode.querySelector('input').focus();
      pill.parentNode.removeChild(pill);
    });

    containerDiv.insertAdjacentElement("afterbegin", span);
    containerDiv.insertAdjacentElement("beforeend", button);

    if (!config.isValid) {
      containerDiv.classList.add("emails-input__root__pill--invalid");
    }

    containerDiv.addEventListener('click', (e) => e.stopPropagation());

    return containerDiv;
  }

  function Input(store) {
    const input = document.createElement("input");
    input.className = "emails-input__root__input";
    input.placeholder = "add more people...";

    input.addEventListener("blur", e => {
      const node = e.currentTarget;
      handleInputValue({ node: node, store: store, value: node.value });
      e.currentTarget.value = "";
    });

    input.addEventListener("keyup", e => {
      const node = e.currentTarget;

      /** Handles pressing ',' or 'Enter' */
      if ([188, 13].indexOf(e.keyCode) !== -1) {
        handleInputValue({ node: node, store: store, value: node.value });
        node.value = "";
      }

      /**
       * Handles pressing backspace to delete items.
       * It assumes the user is deleting the last item added
       * since the input is in the end of the list.
       **/
      if (e.keyCode === 8 && node.value === "") {
        const pills = input.parentNode.querySelectorAll("div");

        if (pills.length) {
          node.parentNode.removeChild(pills[pills.length - 1]);
          store.pop();
        }
      }
    });

    input.addEventListener("paste", e => {
      const node = e.currentTarget;

      /** IE 11 relies on window.clipboardData to get the pasted text */
      const value = (e.clipboardData || window.clipboardData).getData("text");
      handleInputValue({ node: node, store: store, value: value });
      node.value = "";

      /**
       * Prevents the browser from pasting the content
       * since this event happens "before" the content
       * is set in the "value" attribute
       **/
      e.preventDefault();
    });

    input.addEventListener('click', (e) => e.stopPropagation());

    return input;
  }

  function EmailsInput(container) {
    const subscriptions = [];
    const emailsStore = [];
    const rootDiv = document.createElement("div");

    rootDiv.className = "emails-input__root";

    /**
     * Intercepts push, pop and splice to trigger notifications to any subscriptions.
     * Notice that the subscription always receives a property `data` with an array
     * of values.
     **/
    emailsStore.push = item => {
      subscriptions.forEach(cb => {
        cb({ event: "ADDITION", data: [{ ...item }] });
      });

      return Array.prototype.push.call(emailsStore, item);
    };

    emailsStore.pop = () => {
      const itemBeingRemoved = Array.prototype.pop.call(emailsStore);
      subscriptions.forEach(cb => {
        cb({ event: "REMOVAL", data: [{ ...itemBeingRemoved }] });
      });

      return itemBeingRemoved;
    };

    emailsStore.splice = (...args) => {
      const itemsBeingRemoved = Array.prototype.splice.call(
        emailsStore,
        ...args
      );

      subscriptions.forEach(cb => {
        cb({
          event: "REMOVAL",
          data: itemsBeingRemoved.map(item => ({ ...item }))
        });
      });

      return itemsBeingRemoved;
    };

    const input = Input(emailsStore);

    rootDiv.addEventListener('click', () => input.focus());
    rootDiv.insertAdjacentElement("beforeend", input);
    container.insertAdjacentElement("beforeend", rootDiv);

    /**
     * Extends the root component with our API ...
     * If our component had to be removed from the DOM,
     * a method to cleanup these methods shuold be created.
     **/
    container.getAllEmails = function getAllEmails() {
      return emailsStore.concat([]);
    };

    container.addNewEmail = function addNewEmail() {
      handleInputValue({
        node: input,
        store: emailsStore,
        value: generateRandomEmail()
      });
    };

    /**
     * This method implements the observable pattern in which
     * returns the function to `unsubscribe` when we're done
     * listening the changes to the store.
     **/
    container.subscribe = function subscribe(cb) {
      if (typeof cb === "function") {
        subscriptions.push(cb);
      }

      return function unsubscribe() {
        const idx = subscriptions.indexOf(cb);
        if (idx !== -1) {
          subscriptions.splice(idx, 1);
        }
      };
    };

    container.replaceAllEmails = function replaceAllEmails() {
      /** Removes all pills */
      Array.prototype.slice
        .call(rootDiv.querySelectorAll("div.emails-input__root__pill"), 0)
        .forEach(function(node) {
          node.parentNode.removeChild(node);
        });

      /**
       * Removes the items from the store,
       * using splice to trigger a notification on subscriptions
       **/
      const numberExistingEmails = emailsStore.length;
      emailsStore.splice(0, emailsStore.length);

      /**
       * Re-uses the addNewEmail() API to trigger notifications on subscriptions
       */
      for (let counter = 0; counter < numberExistingEmails; counter += 1) {
        container.addNewEmail();
      }
    };
  }

  /** Exposes the method to generate EmailsInput instances */
  window.EmailsInput = EmailsInput;
})(window, document);
