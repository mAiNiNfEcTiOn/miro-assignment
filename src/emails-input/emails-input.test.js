import { fireEvent, waitFor, createEvent } from "@testing-library/dom";

import "@testing-library/jest-dom/extend-expect";

import "./emails-input";

describe("emails-input", () => {
  let div;

  beforeEach(() => {
    div = document.createElement("div");
    div.id = "test";
    document.body.appendChild(div);
  });

  afterEach(() => {
    // Cleanup
    div.parentNode.removeChild(div);
  });

  it("matches the expected initial rendering", async () => {
    window.EmailsInput(div);

    await waitFor(() =>
      expect(document.querySelector(".emails-input__root__input")).toBeVisible()
    );
    expect(document.body).toMatchSnapshot();
  });

  it("adds a pill/block when inserts something and presses enter, comma or blurs out of the input. Marks it as valid or invalid based on being or not a valid e-mail", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );

    input.value = "test";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });

    /** Enter */
    await waitFor(() =>
      expect(document.querySelector(".emails-input__root__pill")).toBeVisible()
    );
    const firstPill = document.querySelector(".emails-input__root__pill");
    expect(firstPill).toHaveTextContent("test");
    expect(firstPill).toHaveClass("emails-input__root__pill--invalid");

    /** Comma */
    input.value = "valid-email@miro.com";
    fireEvent.keyUp(input, { key: ",", keyCode: 188 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(2)
    );

    const secondPill = document.querySelector(
      ".emails-input__root__pill:nth-child(2)"
    );
    expect(secondPill).toHaveTextContent("valid-email@miro.com");
    expect(secondPill).not.toHaveClass("emails-input__root__pill--invalid");

    /** Blur */
    input.value = "another-valid-email@miro.com";
    fireEvent.blur(input);
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(3)
    );

    const thirdPill = document.querySelector(
      ".emails-input__root__pill:nth-child(3)"
    );
    expect(thirdPill).toHaveTextContent("another-valid-email@miro.com");
    expect(thirdPill).not.toHaveClass("emails-input__root__pill--invalid");

    expect(document.body).toMatchSnapshot();
  });

  it("ignores empty entries", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );
    input.value = "";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(0)
    );

    input.value = ",,,,,,";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(0)
    );
  });

  it("ignores attempt to insert the same value twice", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );
    input.value = "test";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(1)
    );

    input.value = "test";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(1)
    );
  });

  it("removes a pill/block ONLY when the input element is empty and the user presses backspace", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );

    input.value = "test";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(document.querySelector(".emails-input__root__pill")).toBeVisible()
    );

    input.value = "some value";
    fireEvent.keyUp(input, { key: "Backspace", keyCode: 8 });
    await waitFor(() =>
      expect(document.querySelector(".emails-input__root__pill")).toBeVisible()
    );

    input.value = "";
    fireEvent.keyUp(input, { key: "Backspace", keyCode: 8 });
    await waitFor(() =>
      expect(document.querySelector(".emails-input__root__pill")).toBe(null)
    );
  });

  it("removes a pill/block when clicking the remove button in it", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );

    input.value = "test";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(document.querySelector(".emails-input__root__pill")).toBeVisible()
    );
    const pill = document.querySelector(".emails-input__root__pill");
    const removeButton = pill.querySelector(
      ".emails-input__root__pill__button"
    );

    fireEvent.click(removeButton);
    await waitFor(() =>
      expect(document.querySelector(".emails-input__root__pill")).toBe(null)
    );
  });

  it("adds random e-mails when calling the addNewEmail method added to the target container", async () => {
    window.EmailsInput(div);

    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(0)
    );

    div.addNewEmail();
    div.addNewEmail();

    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(2)
    );
  });

  it("retrieves the list of ALL e-mails added - invalid and valid - when calling the getAllEmails method added to the target container", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );
    input.value = "test";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });

    input.value = "valid@email.dom";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(2)
    );

    expect(div.getAllEmails()).toMatchObject([
      {
        isValid: false,
        value: "test"
      },
      {
        isValid: true,
        value: "valid@email.dom"
      }
    ]);
  });

  it("replaces the existing list of e-mails with new ones when calling the replaceAllEmails method added to the target container", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );

    div.addNewEmail();
    div.addNewEmail();
    div.addNewEmail();
    div.addNewEmail();

    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(4)
    );
    const initialPillsTexts = Array.from(document.querySelectorAll(".emails-input__root__pill")).map((node) => node.textContent);

    div.replaceAllEmails();
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(4)
    );
    const finalPillsTexts = Array.from(document.querySelectorAll(".emails-input__root__pill")).map((node) => node.textContent);

    expect(initialPillsTexts).not.toMatchObject(finalPillsTexts);
  });

  it("receives additions and removals from the list of e-mails when subscribed and stops receiving after unsubscribing", async () => {
    const myFn = jest.fn();
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );
    input.value = "test";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(1)
    );

    expect(myFn).toHaveBeenCalledTimes(0);

    // Subscribes
    const unsubscribeFn = div.subscribe(entries => myFn(entries));

    // Adds 1 valid e-mail
    input.value = "valid@email.dom";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(2)
    );

    expect(myFn).toHaveBeenCalledTimes(1);
    expect(myFn).toHaveBeenCalledWith({
      event: "ADDITION",
      data: [{ isValid: true, value: "valid@email.dom" }]
    });
    myFn.mockClear();

    // Adds another valid e-mail
    input.value = "my@email.dom";
    fireEvent.keyUp(input, { key: ",", keyCode: 188 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(3)
    );

    expect(myFn).toHaveBeenCalledTimes(1);
    expect(myFn).toHaveBeenCalledWith({
      event: "ADDITION",
      data: [{ isValid: true, value: "my@email.dom" }]
    });
    myFn.mockClear();

    // Removes one e-mail via backspace
    fireEvent.keyUp(input, { key: "Backspace", keyCode: 8 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(2)
    );

    expect(myFn).toHaveBeenCalledTimes(1);
    expect(myFn).toHaveBeenCalledWith({
      event: "REMOVAL",
      data: [{ isValid: true, value: "my@email.dom" }]
    });
    myFn.mockClear();

    // Removes another e-mail via clicking in the button
    const pillRemoveButton = document.querySelector(".emails-input__root__pill:nth-child(2) button");
    fireEvent.click(pillRemoveButton);
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(1)
    );

    expect(myFn).toHaveBeenCalledTimes(1);
    expect(myFn).toHaveBeenCalledWith({
      event: "REMOVAL",
      data: [{ isValid: true, value: "valid@email.dom" }]
    });
    myFn.mockClear();

    // Unsubscribes
    unsubscribeFn();

    input.value = "another-valid@email.miro.com";
    fireEvent.keyUp(input, { key: "Enter", keyCode: 13 });
    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(2)
    );

    expect(myFn).toHaveBeenCalledTimes(0);
  });

  it("parse values pasted into proper pills", async () => {
    window.EmailsInput(div);

    const input = await waitFor(() =>
      document.querySelector(".emails-input__root__input")
    );

    const pastedData = "multiple,items, in the same, entry, being pasted";

    /**
     * This needed due to a limitation in JSDom's (non-?)implementation of the ClipboardEvent
     * @see {@link https://stackoverflow.com/a/58452222}
     **/
    const pasteEventData = {
      clipboardData: {
        types: ["text/plain", "text/html"],
        getData: type => pastedData,
      }
    };
    const pasteEvent = createEvent.paste(input, pasteEventData);
    pasteEvent.clipboardData = pasteEventData.clipboardData;
    fireEvent(input, pasteEvent);

    await waitFor(() =>
      expect(
        document.querySelectorAll(".emails-input__root__pill").length
      ).toBe(5)
    );
    
    const pillsTexts = Array.from(document.querySelectorAll(".emails-input__root__pill")).map((node) => node.textContent);
    expect(pillsTexts).toMatchObject(['multiple', 'items', 'in the same', 'entry', 'being pasted'])
  });
});
