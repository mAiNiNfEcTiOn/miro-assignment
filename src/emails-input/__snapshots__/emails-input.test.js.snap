// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emails-input adds a pill/block when inserts something and presses enter, comma or blurs out of the input. Marks it as valid or invalid based on being or not a valid e-mail 1`] = `
<body>
  <div
    id="test"
  >
    <div
      class="emails-input__root"
    >
      <div
        class="emails-input__root__pill emails-input__root__pill--invalid"
      >
        <span>
          test
        </span>
        <button
          class="emails-input__root__pill__button"
          type="button"
        >
          <svg
            fill="none"
            focusable="false"
            height="8"
            tabindex="-1"
            viewBox="0 0 8 8"
            width="8"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 0.8L7.2 0L4 3.2L0.8 0L0 0.8L3.2 4L0 7.2L0.8 8L4 4.8L7.2 8L8 7.2L4.8 4L8 0.8Z"
              fill="#050038"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="emails-input__root__pill"
      >
        <span>
          valid-email@miro.com
        </span>
        <button
          class="emails-input__root__pill__button"
          type="button"
        >
          <svg
            fill="none"
            focusable="false"
            height="8"
            tabindex="-1"
            viewBox="0 0 8 8"
            width="8"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 0.8L7.2 0L4 3.2L0.8 0L0 0.8L3.2 4L0 7.2L0.8 8L4 4.8L7.2 8L8 7.2L4.8 4L8 0.8Z"
              fill="#050038"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="emails-input__root__pill"
      >
        <span>
          another-valid-email@miro.com
        </span>
        <button
          class="emails-input__root__pill__button"
          type="button"
        >
          <svg
            fill="none"
            focusable="false"
            height="8"
            tabindex="-1"
            viewBox="0 0 8 8"
            width="8"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 0.8L7.2 0L4 3.2L0.8 0L0 0.8L3.2 4L0 7.2L0.8 8L4 4.8L7.2 8L8 7.2L4.8 4L8 0.8Z"
              fill="#050038"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <input
        class="emails-input__root__input"
        placeholder="add more people..."
      />
    </div>
  </div>
</body>
`;

exports[`emails-input matches the expected initial rendering 1`] = `
<body>
  <div
    id="test"
  >
    <div
      class="emails-input__root"
    >
      <input
        class="emails-input__root__input"
        placeholder="add more people..."
      />
    </div>
  </div>
</body>
`;
